### ISSUE TITLE
FIM DB synchronization between managers

### ISSUE THREAD

**Issue by jlruizmlg** [wazuh/wazuh#2404]
|Wazuh version|Install type|Install method|Platform|
|---|---|---|---|
| v3.8.2 | Manager | Packages/Sources | Any |

Hello team,
in a cluster environment, whenever agents move from node to node it generates file integrity checksum changed alerts for changes already reported.

We need some mechanism to synchronize DBs among different Wazuh manager nodes.

Regards.

**Comment by jlruizmlg** [comment]
Any new ETA about this issue?

Thanks.

**Comment by mgmacias95** [comment]
Hello @jlruizmlg,

We have done a full rework of the Wazuh cluster to improve the performance of the synchronization process. This rework will allow us to share larger files among cluster nodes including FIM databases. This rework is the first step towards synchronizing FIM and PM databases among cluster nodes, but we don't still have an ETA for it.

Best regards,
Marta

**Comment by bah07** [comment]
## Problem:

An agent connected to manager001 changes and starts to report to manager002, this change produces inconsistencies in the content of the databases. Since the events generated in the agent produce updates in the content of the database, in case the agent changes to report to another manager the changes that occurred while reporting to manager001 are not made on the database of manager002. This produces several inconveniences:

- False positives are reported. In the case that an agent returns to report to a manager that has not reported for a while (this contains an obsolete database of the agent) the files that have been modified in that time interval will be reported as false positives since these changes were already reflected in the second manager.

- Obsolete/wrong information is obtained in the API requests about the information stored in the DB. Since the content is not updated until the agent is restarted, the requests from the API return wrong data.

## Solution:

- Integrity check of the manager's DB, every so often the database status is checked and updated if necessary.

- The agent reports the new state of the file that generates the event and also the alert that the manager has to show. Avoiding this way to report obsolete information stored in the manager.

- The status of the file sent by the agent is reported with all the attributes regardless of the configured choices.

## Testing:

- [ ] Check that there are no false positives in cluster environments in continuous changes with load balancers of FIM events.

- [ ] Verify that integrity checks update databases when necessary.

- [ ] Check that there are no alerts in the integrity cycle between agent and manager.

**Comment by bah07** [comment]
Proposed solutions to integrity checks:

1st Solution. Check integrity by blocks.

Divide the database in the agent in a number of blocks by a value defined in the configuration. The database entries are ordered alphabetically by the path and each entry is assigned to its corresponding block and the integrity hash is calculated for each set of entries.

For example:

Configuration: 

```
<directories check_all="yes">/etc/X11/app-defaults</directories>
<blocks>10</blocks>
```

### BLOCK 1
```
File: '/etc/X11/app-defaults/Bitmap'
File: '/etc/X11/app-defaults/Bitmap-color'
File: '/etc/X11/app-defaults/Bitmap-nocase'
File: '/etc/X11/app-defaults/Clock-color'

HASH: XXXXXXXXXX1
```

### BLOCK 2
```
File: '/etc/X11/app-defaults/Editres'
File: '/etc/X11/app-defaults/Editres-color'
File: '/etc/X11/app-defaults/KOI8RXTerm'
File: '/etc/X11/app-defaults/KOI8RXTerm-color'

HASH: XXXXXXXXXX2
```

### BLOCK 3
```
File: '/etc/X11/app-defaults/UXTerm'
File: '/etc/X11/app-defaults/UXTerm-color'
File: '/etc/X11/app-defaults/Viewres'
File: '/etc/X11/app-defaults/Viewres-color'

HASH: XXXXXXXXXX3
```

### BLOCK 4
```
File: '/etc/X11/app-defaults/XCalc'
File: '/etc/X11/app-defaults/XCalc-color'
File: '/etc/X11/app-defaults/XClipboard'
File: '/etc/X11/app-defaults/XClock'

HASH: XXXXXXXXXX4
```

### BLOCK 5
```
File: '/etc/X11/app-defaults/XClock-color'
File: '/etc/X11/app-defaults/XConsole'
File: '/etc/X11/app-defaults/XFontSel'
File: '/etc/X11/app-defaults/XLoad'

HASH: XXXXXXXXXX5
```

### BLOCK 6
```
File: '/etc/X11/app-defaults/XLogo'
File: '/etc/X11/app-defaults/XLogo-color'
File: '/etc/X11/app-defaults/XMore'
File: '/etc/X11/app-defaults/XSm'

HASH: XXXXXXXXXX6
```

### BLOCK 7
```
File: '/etc/X11/app-defaults/XTerm'
File: '/etc/X11/app-defaults/XTerm-color'
File: '/etc/X11/app-defaults/Xditview'
File: '/etc/X11/app-defaults/Xditview-chrtr'

HASH: XXXXXXXXXX7
```

### BLOCK 8
```
File: '/etc/X11/app-defaults/Xedit'
File: '/etc/X11/app-defaults/Xedit-color'
File: '/etc/X11/app-defaults/Xfd'
File: '/etc/X11/app-defaults/Xgc'

HASH: XXXXXXXXXX8
```

### BLOCK 9
```
File: '/etc/X11/app-defaults/Xgc-color'
File: '/etc/X11/app-defaults/Xmag'
File: '/etc/X11/app-defaults/Xman'
File: '/etc/X11/app-defaults/Xmessage'

HASH: XXXXXXXXXX9
```

### BLOCK 10
```
File: '/etc/X11/app-defaults/Xmessage-color'
File: '/etc/X11/app-defaults/Xvidtune'

HASH: XXXXXXXXX10
```

```
HASH COMPLETE: 12345678910

HASH1: XXXXXXXXXX1
HASH2: XXXXXXXXXX2
HASH3: XXXXXXXXXX3
HASH4: XXXXXXXXXX4
HASH5: XXXXXXXXXX5
HASH6: XXXXXXXXXX6
HASH7: XXXXXXXXXX7
HASH8: XXXXXXXXXX8
HASH9: XXXXXXXXXX9
```

The manager checks each block and only requests the information of the blocks that do not match.

Problem: The algorithm makes a path alphabetically sorting the entries, in case of inserting a new entry does not only change the block where the entry falls but in all the blocks that remain after, as they move all the entries in one position. When a different hash is generated in all the blocks when they are checked in the manager, all the mentioned blocks are sent, even if only one element has changed.

To solve this problem we can use the hash table to make the division by blocks and that does not depend on the alphabetical order. 

| ID    | KEY                                       | DATA                |
|-------|-------------------------------------------|---------------------|
| 0     |  '/etc/X11/app-defaults/Bitmap'           |       ...           |
| 1     |  '/etc/X11/app-defaults/Bitmap-color'     |       ...           |
| 2     |  '/etc/X11/app-defaults/Bitmap-nocase'    |       ...           |
| 3     |  '/etc/X11/app-defaults/Clock-color'      |       ...           |
| 4     |  '/etc/X11/app-defaults/Editres'          |       ...           |
| 5     |  '/etc/X11/app-defaults/Editres-color'    |       ...           |
| 6     |  '/etc/X11/app-defaults/KOI8RXTerm'       |       ...           |
| 7     |  '/etc/X11/app-defaults/KOI8RXTerm-color' |       ...           |
| 8     |  '/etc/X11/app-defaults/UXTerm'           |       ...           |
| 9     |  '/etc/X11/app-defaults/UXTerm-color'     |       ...           |
| 10    |  '/etc/X11/app-defaults/Viewres'          |       ...           |
| 11    |  '/etc/X11/app-defaults/Viewres-color'    |       ...           |
| 12    |  '/etc/X11/app-defaults/XCalc'            |       ...           |
| 13    |  '/etc/X11/app-defaults/XCalc-c           |       ...           |
| 14    |  '/etc/X11/app-defaults/XClipbo           |       ...           |
| 15    |  '/etc/X11/app-defaults/XClock'           |       ...           |

Bearing in mind that the KEY field can be a list of elements (collisions in the hash table). You obtain the elements of a block, for example divide in 4 blocks:

BLOCK 1 {1, 2, 3, 4} 
BLOCK 2 {5, 6, 7, 8} 
BLOCK 3 {9, 10, 11, 12} 
BLOCK 4 {13, 14, 15, 16} 

Elements are alphabetically ordered to calculate their integrity. In addition, a hash is calculated with the id's of each block, which will be the hash_block. Both the data and the hash block corresponding to each entry are stored in the manager's database:

| KEY                                       | DATA                | HASH BLOCK       |
|-------------------------------------------|---------------------|------------------|
|  '/etc/X11/app-defaults/Bitmap'           |                     | {1, 2, 3, 4}     |
|  '/etc/X11/app-defaults/Bitmap-color'     |                     | {1, 2, 3, 4}     |
|  '/etc/X11/app-defaults/Bitmap-nocase'    |                     | {1, 2, 3, 4}     |
|  '/etc/X11/app-defaults/Clock-color'      |                     | {1, 2, 3, 4}     |
|  '/etc/X11/app-defaults/Editres'          |                     | {5, 6, 7, 8}     |
|  '/etc/X11/app-defaults/Editres-color'    |                     | {5, 6, 7, 8}     |
|  '/etc/X11/app-defaults/KOI8RXTerm'       |                     | {5, 6, 7, 8}     |
|  '/etc/X11/app-defaults/KOI8RXTerm-color' |                     | {5, 6, 7, 8}     |
|  '/etc/X11/app-defaults/UXTerm'           |                     | {9, 10, 11, 12}  |
|  '/etc/X11/app-defaults/UXTerm-color'     |                     | {9, 10, 11, 12}  |
|  '/etc/X11/app-defaults/Viewres'          |                     | {9, 10, 11, 12}  |
|  '/etc/X11/app-defaults/Viewres-color'    |                     | {9, 10, 11, 12}  |
|  '/etc/X11/app-defaults/XCalc'            |                     | {13, 14, 15, 16} |
|  '/etc/X11/app-defaults/XCalc-c           |                     | {13, 14, 15, 16} |
|  '/etc/X11/app-defaults/XClipbo           |                     | {13, 14, 15, 16} |
|  '/etc/X11/app-defaults/XClock'           |                     | {13, 14, 15, 16} |

So the manager can calculate the integrity of each block in the same way as the agent...

Use case:
```
Agent                                                            Manager
|                                                                |
| Create a hash table                                              |
| Generate hash_block                                            |
|                                                                |
|                                                                |
| Send integrity checks ---------------------------------------->|
|                                                                |
|                                                                |
|<---------------------------------------------------------------| Request full database
|                                                                |
|                                                                |
|Sending complete database ------------------------------------->|
|                                                                |
|                                                                |
| **The agent changes node and reports                           |
| x elements in another manager.                                 |
| i.e '/etc/X11/app-defaults/UXTerm'                             |
| This manager updates his database and shows                    |
| the corresponding alert. Later, the agent                      |
| reports back to the old manager, having this                    |
| database outdated by the change produced in                    |
| the other manager.                                             |
|                                                                |
|                                                                |
| Send integrity checks ---------------------------------------->|
|                                                                |
|                                                                |
|                                                                | Integrity check of block '{9, 10, 11, 12}' failed
|<---------------------------------------------------------------| Request block '{9, 10, 11, 12}'
|                                                                |
|                                                                |
```

This methodology has one drawback and that is that it does not have a good balance between the number of Agent-Manager communications.

Ideally we want to get the balance between the number of control messages needed to make the integrity check and the number of messages that are sent when the integrity check of a part of the database fails and is updated.


By creating 3 levels of blocks we can specify in a more granular way the incorrect integrity check, making two more checks.

We reduce the problem to a hash table of 16 elements (we estimate that in each row we can nest up to 200 entries without loss of performance by linear searches). We reduce to the maximum the granularity of the first level, the number of blocks of level 0 is equal to the number of tuples of the hash table.

```
|   0   |   1   |   2   |   3   |   4   |   5   |   6   |   7   |   8   |   9   |   10   |   11   |   12   |   13   |   14   |   15   |
|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|--------|--------|--------|--------|--------|--------|
| HL0B1 | HL0B2 | HL0B3 | HL0B4 | HL0B5 | HL0B6 | HL0B7 | HL0B8 | HL0B9 |HL0B10 | HL0B11 | HL0B12 | HL0B13 | HL0B14 | HL0B15 | HL0B16 | Blocks level 0
|---------------|---------------|---------------|---------------|---------------|-----------------|-----------------|-----------------|
|      HL1B1    |      HL1B2    |      HL1B3    |      HL1B4    |      HL1B5    |      HL1B6      |      HL1B7      |      HL1B8      | Blocks level 1
|-------------------------------|-------------------------------|---------------------------------|-----------------------------------|
|             HL2B1             |               HL2B2           |               HL2B3             |               HL2B4               | Blocks level 2
```

In order to check the integrity, a message will be sent to the manager where the block and the integrity hash of that block will be coded.
```
Level: 2
Block: 3
Hash: 3f5f8940080f1f17ea3026480ca214153bf94041
```

The SQL table in manager side has 3 columns to store each hash_block

|        Path                               |        Data         |        L0        |        L1        |        L2        |
|-------------------------------------------|---------------------|------------------|------------------|------------------|
|  '/etc/X11/app-defaults/Bitmap'           |         ...         |      HL0B1       |      HL1B1       |      HL2B1       |
|  '/etc/X11/app-defaults/Bitmap-color'     |         ...         |      HL0B2       |      HL1B1       |      HL2B1       |
|  '/etc/X11/app-defaults/Bitmap-nocase'    |         ...         |      HL0B3       |      HL1B2       |      HL2B1       |
|  '/etc/X11/app-defaults/Clock-color'      |         ...         |      HL0B4       |      HL1B2       |      HL2B1       |
|  '/etc/X11/app-defaults/Editres'          |         ...         |      HL0B5       |      HL1B3       |      HL2B2       |
|  '/etc/X11/app-defaults/Editres-color'    |         ...         |      HL0B6       |      HL1B3       |      HL2B2       |
|  '/etc/X11/app-defaults/KOI8RXTerm'       |         ...         |      HL0B7       |      HL1B4       |      HL2B2       |
|  '/etc/X11/app-defaults/KOI8RXTerm-color' |         ...         |      HL0B8       |      HL1B4       |      HL2B2       |
|  '/etc/X11/app-defaults/UXTerm'           |         ...         |      HL0B9       |      HL1B5       |      HL2B3       |
|  '/etc/X11/app-defaults/UXTerm-color'     |         ...         |      HL0B10      |      HL1B5       |      HL2B3       |
|  '/etc/X11/app-defaults/Viewres'          |         ...         |      HL0B11      |      HL1B6       |      HL2B3       |
|  '/etc/X11/app-defaults/Viewres-color'    |         ...         |      HL0B12      |      HL1B6       |      HL2B3       |
|  '/etc/X11/app-defaults/XCalc'            |         ...         |      HL0B13      |      HL1B7       |      HL2B4       |
|  '/etc/X11/app-defaults/XCalc-c           |         ...         |      HL0B14      |      HL1B7       |      HL2B4       |
|  '/etc/X11/app-defaults/XClipbo           |         ...         |      HL0B15      |      HL1B8       |      HL2B4       |
|  '/etc/X11/app-defaults/XClock'           |         ...         |      HL0B16      |      HL1B8       |      HL2B4       |

To calculate the integrity of block HL2B1. We need:

|        Path                               |        Data         |        L0        |        L1        |        L2        |
|-------------------------------------------|---------------------|------------------|------------------|------------------|
|  '/etc/X11/app-defaults/UXTerm'           |         ...         |      HL0B9       |      HL1B5       |      HL2B3       |
|  '/etc/X11/app-defaults/UXTerm-color'     |         ...         |      HL0B10      |      HL1B5       |      HL2B3       |
|  '/etc/X11/app-defaults/Viewres'          |         ...         |      HL0B11      |      HL1B6       |      HL2B3       |
|  '/etc/X11/app-defaults/Viewres-color'    |         ...         |      HL0B12      |      HL1B6       |      HL2B3       |

array = SELECT data FROM FIM_ENTRY WHERE L2=HL2B3

hash = getHash(array)

If the hash is different, the manager will send a hash failure signal of level 2 block 3. Then, the agent will return in a message the integrity hashes of the corresponding sub-blocks.
```
Level: 1
Block: 5
Hash: 76b57f62652bddad6f0932d4394093bf95c99457

Level: 1
Block: 6
Hash: f26f2e19c86a6237dea626edee1977e217244946
```

Until the last block is reduced, finally the agent will send directly the updated content of the failed blocks.




2nd Solution. Generate the blocks by directories. The number of elements in each block depends on the number of files in a directory. We limit the maximum number of entries that will contain a block but always completing a directory. If a directory has more than n elements these are separated in sub-blocks as in the first case, we will have the same problem, but we reduce the case to a specific directory that exceeds the number of entries per block.

Here we have two alternatives

- We can generate hashes only from files that contain a directory, with this what we do is that the hash of a directory does not depend on the hash of its subdirectories. We send a message with all the hashes of all the directories, here the problem is in the size of the generated message.

- Generate hashes of files and directories, in this case, we have to send the integrity hashes for each directory. In case of failure, we send the hashes of the integrity of those subdirectories, until we find the directory whose hash is wrong.

**Comment by Cerv1** [comment]
### Development update

We have changed the number of `fim_entry_data` hash tables. Previously, we thought that we need 3 `fim_entry_data` tables (one for each FIM mode), but actually, we managed to merge all of them into only one. The main reason for this merge was because now it’s much simpler to obtain the integrity hashes due to its division into blocks. 

To distinguish each entry we have added a `mode` attribute in the data structure that indicates which mode has inserted that entry. This `mode` can be one of the following:
* `FIM_SCHEDULED` (0)
* `FIM_REALTIME` (1)
* `FIM_WHODATA` (2)

```c
typedef struct fim_entry_data {
   // Checksum attributes
   unsigned int size;
   unsigned int perm;
   unsigned int uid;
   unsigned int gid;
   char * user_name;
   char * group_name;
   unsigned int mtime;
   long unsigned int inode;
   os_md5 hash_md5;
   os_sha1 hash_sha1;
   os_sha256 hash_sha256;
   // Options
   int mode;
   int options;
} fim_entry_data;
```
Best regards,
Cerv1.

**Comment by vikman90** [comment]
## Sychronize hash table with a database

### Database table design

The database needs to store some new items:

1. Which block each entry belongs.
2. Hash of each entry, for performance reasons.
3. Hash of the blocks (to be discussed).

#### Table `fim_entry`

We will add three columns: 

- `hash`, that stores the hash of each entry.
- `block_l1` and `block_l2`, that will indicate the block which each file belongs to.

|Column|Type|Description|
|---|---|---|
|`file`|Text|Primary key.|
|`type`|Text|Not null, value `file` or `registry`.|
|`date`|Integer|Not null, default is current time.|
|`changes`|Integer|Not null, default is `1`.|
|`size`|Integer||
|`perm`|Text||
|`uid`|Text||
|`gid`|Text||
|`md5`|Text||
|`sha1`|Text||
|`uname`|Text||
|`gname`|Text||
|`mtime`|Integer||
|`inode`|Integer||
|`sha256`|Text||
|`attributes`|Integer|Default is `0`.|
|`symbolic_path`|Text||
|`hash`|Text|
|`block_l1`|Integer||
|`block_l2`|Integer||

#### Table `block_hash`

This table stores the accumulated hash of each entry. It remains to be discussed if this is necessary.

|Column|Type|Description|
|---|---|---|
|`component`|Text|Primary key.|
|`level`|Integer|Primary key.|
|`blockno`|Integer|Primary key.|
|`hash`|Text|Not null.|

### Example of integrity synchronization protocol

#### Agent's payload

The agent will send a message that may contain the hash of one or many blocks of the same level.

For instance:

```json
{
  "type": "integrity_check",
  "level": 2,
  "ofset": 5,
  "blocks": [
    "01234567",
    "...",
    "89ABCDEF"
  ]
}
```

#### Analysisd - Wazuh DB protocol

```
[Analysisd → Wazuh DB] fim integrity check { "block": 5, "level": 2, "hash": "01234567" }
[Wazuh DB → Analysisd] ok | ok 76543210
[Analysisd → Agent]    block_expand { "block": 5, "level": 2 }
```

### How to calculate hash in the manager side

Two options:

1. Each `INSERT` or `REPLACE` query shall produce a file hash and spread towards the upper blocks.

Efficiency:

> n level-2 blocks ( ( hash of 0..∛n sibling files ) + ( hash of hash of 0..∛n level-1 sibling blocks ) + level-2 hash )
> n • ( ∛n÷2 + ∛n÷2 ) = n • ∛n = **O(∛n⁴)**

2. Each integrity check query shall gather the corresponding files' hash and produce a large-grained block hash.

Efficiency:

> ∛n level-2 blocks • ( ∛n level-1 blocks • ( ∛n files ) )
> +
> ∛n² level-1 blocks • ( ∛n files )
> +
> n files

> O(3n) ≈ **O(n)**

Let's compare both algoritms:

<img width="420" alt="compare" src="https://user-images.githubusercontent.com/10536251/62884600-3f0a4780-bd37-11e9-88de-a5924425c365.png">

According to this chart, option 2 is more efficient so the database will perform the hash of each block when executing an integrity check query.

**Comment by vikman90** [comment]
## New design stage

### Feedback

1. We do need a block hash structure that may remain in the time. In the agent side, FIM will create a state snapshot and will synchronize that snapshot.
2. A level-0 block is not a file but a hash table cell. In other words, it's a set of files. So we need to store the level-0 block number in the table `fim_entry`.
3. The manager needs to know if a bunch of states is the last one so it can delete the remaining files. This can happen if the list of files in the agent decreases drastically.
4. The manager needs to synchronize the version of its state with the agent's state. This means that the agent shall append a "scan ID" so the manager will recreate the state again (if the agent's message ID is greater) or drop the message (if the agent's message ID is lower).
5. `block_l1` depends on `block_l0`, and `block_l2` depends on `block_l1`. Having these columns together in the same table violates the [third normal form](https://en.wikipedia.org/wiki/Third_normal_form).

### New table design

#### Table `fim_entry`

New columns:

|Column|Type|Description|
|---|---|---|
|`hash`|Text|Hash of the entire file (path + attributes).|
|`block_l0`|Integer|Level-0 block which this file belongs to.|

#### Table `block_hash`

|Column|Type|Constraint|Description|
|---|---|---|---|
|`component`|Text|Primary key.|Name of the component (`fim`, `sca`, etc.)|
|`level`|Integer|Primary key. Value `0`, `1` or `2`.|Number of the level.|
|`blockno`|Integer|Primary key.|Number of block in that level.|
|`hash`|Text|Not null.|Accumulated hash of the block.|

#### Table `integrity_level`

This colum will store each block's parent.

|Column|Type|Constraint|Description|
|---|---|---|---|
|`component`|Text|Primary key.|Name of the component (`fim`, `sca`, etc.)|
|`level`|Integer|Primary key. Value `0` or `1`.|Number of the child level.|
|`block`|Integer|Primary key.|Number of the child block.|
|`parent`|Integer||Number of the parent block.|

#### Table `integrity_version`

This colum will store each component's integrity version.

*TBD: We could store this in the table `metadata`.*

|Column|Type|Constraint|Description|
|---|---|---|---|
|`component`|Text|Primary key.|Name of the component (`fim`, `sca`, etc.)|
|`id`|Integer|Primary key.|ID of the currently stored block hash.|

### Synchronization protocol

#### Message from agent to the manager

```json
{
  "type": "integrity_check",
  "id": 1565721013,
  "level": 2,
  "ofset": 5,
  "blocks": [
    "01234567",
    "...",
    "89ABCDEF"
  ],
  "more_blocks": false
}
```

#### Message from Analysisd to Wazuh DB

```
agent 001 fim integrity_check { "id": 1565721013, "level": 2, "block": 5, "hash": "01234567" }
```

### Regeneration of hash snapshot

#### Condition

Wazuh DB will automatically regenerate the hash state for a component when it receives an `integrity_check` query with an ID number greater than the currently stored version ID for that component.

#### Algorithm

1. For each level-0 block: 
  1.1. Set an empty SHA-1 context.
  1.2. Iterate down the list of files belonging to that block, ordered by `file`, and update the SHA-1 hash with each file's hash.
  1.3. Store the hash digest into the table `block_hash` for that block.
2. For each level-1 block:
  2.1. Set up an empty SHA-1 context.
  2.2. Iterate down the list of level-0 blocks belonging to that level-1 block, ordered, and update the SHA-1 hash with each block's hash.
  2.3. Store the hash digest into the table `block_hash` for that block.
3. For each level-2 block:
  2.1. Set up an empty SHA-1 context.
  2.2. Iterate down the list of level-1 blocks belonging to that level-2 block, ordered, and update the SHA-1 hash with each block's hash.
  2.3. Store the hash digest into the table `block_hash` for that block.

##### Efficiency of this algorithm

Assuming _n_ level-0 blocks (hash table cells) and _m_ files per cell. The agent regulates the number of files per cell, so we can consider _m_ is proportional to _n_.

1. Step 1 is _O(n)_.
2. Step 2 is _O(∛n²)_.
3. Step 3 is _O(∛n)_.

_O(n) + O(∛n²) + O(∛n)_ is **_O(n)_**:

<img width="479" alt="Plot" src="https://user-images.githubusercontent.com/10536251/62969545-c4f8c200-be0d-11e9-8379-6659c98d3fd3.png">

**Comment by vikman90** [comment]
@jlruizmlg this issue has been implemented in #3319. Let me close it.
